### PR TITLE
Don't use Thread.sleep() when rejecting requests

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Essentially the fix is:

- Keep track of when to wake up and start responding to messages
- If the drop option wasn't specified, then we have to respond to the messages which were sent during the sleeping stage. But the only way to reliably do that is to manually keep track of the message queue.

This could could be simplified further I think..

### CODE REVIEW

File: source/geod24/LocalRest.d [hunk 1]
```diff
@@ -400,34 +400,96 @@ public final class RemoteAPI (API) : API
...
+        import std.range;
+        SysTime sleep_until;
+        bool drop_messages;
+
+        bool isSleeping()
+        {
+            return sleep_until != SysTime.init && Clock.currTime < sleep_until;
+        }
+
+        void handle (T) (T arg)
```
Comment (AndrejMitrovic) [inline source/geod24/LocalRest.d]: Not atomic, should be moved to its own commit.

**Comment by AndrejMitrovic** [comment]
Split it up into atomic commits.

File: source/geod24/LocalRest.d [hunk 2]
```diff
@@ -421,12 +434,11 @@ public final class RemoteAPI (API) : API
                             filter = filter_api;
                         },
                         (Response res) {
-                            scheduler.pending = res;
-                            scheduler.waiting[res.id].c.notify();
+                            handle(res);
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: I think you could just pass `handle!Response, handle!Command` then

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: what

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: Ah you mean `&handle!Response`

Comment (Geod24) [reply source/geod24/LocalRest.d]: Yup, sorry

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: I wanted to do it just now, but then I realized in the second commit we add a bit of logic here.

File: source/geod24/LocalRest.d [hunk 3]
```diff
@@ -400,12 +400,23 @@ public final class RemoteAPI (API) : API
...
+        import std.datetime.systime;
+        import std.range;
         scope node = new Implementation(cargs);
         scheduler = new LocalScheduler;
         scope exc = new Exception("You should never see this exception - please report a bug");
 
         FilterAPI filter;
 
+        bool drop_messages;
+        SysTime sleep_until;
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Can we unify those 3 things under one local `struct` ?

File: source/geod24/LocalRest.d [hunk 4]
```diff
@@ -419,27 +430,59 @@ public final class RemoteAPI (API) : API
...
-                            handle(res);
+                            if (!isSleeping())
+                            {
+                                handle(res);
+                            }
+                            else
+                            {
+                                if (!drop_messages)
+                                    response_queue ~= res;
+                            }
```
Comment (Geod24) [inline source/geod24/LocalRest.d]:
```d
if (!isSleeping())
    handle(res)
else if (!drop_messages)
    response_queue ~= res;
```

File: source/geod24/LocalRest.d [hunk 5]
```diff
@@ -419,27 +430,59 @@ public final class RemoteAPI (API) : API
...
                         },
                         (Command cmd)
                         {
-                            handle(cmd);
+                            if (!isSleeping())
+                            {
+                                handle(cmd);
+                            }
+                            else
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Same

File: source/geod24/LocalRest.d [hunk 6]
```diff
@@ -419,27 +430,59 @@ public final class RemoteAPI (API) : API
...
                         });
+
+                    // now handle any leftover messages after any sleep() call
+                    if (!isSleeping())
+                    {
+                        response_queue.each!(res => handle(res));
+                        response_queue.length = 0;
+
+                        command_queue.each!(res => handle(res));
+                        command_queue.length = 0;
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: There's a problem with this: Messages need to be processed in the order they are received

Comment (AndrejMitrovic) [reply source/geod24/LocalRest.d]: That's a good point.

**Comment by AndrejMitrovic** [comment]
So the only way I've made this work is by using Variant. I'm not sure if there is a better way to do it.

File: source/geod24/LocalRest.d [hunk 7]
```diff
@@ -413,21 +452,39 @@ public final class RemoteAPI (API) : API
...
+                        await_msgs.each!((Variant msg)
+                        {
+                            if (auto cmd = msg.peek!Command)
+                                handle(*cmd);
+                            else if (auto res = msg.peek!Response)
+                                handle(*res);
+                            else assert(0, "Unhandled message: %s".format(msg));
                         });
+
+                        await_msgs.length = 0;
```
Comment (AndrejMitrovic) [inline source/geod24/LocalRest.d:492]:
```suggestion
                        await_msgs.length = 0;
                        assumeSafeAppend(await_msgs);
```

?

File: source/geod24/LocalRest.d [hunk 8]
```diff
@@ -40,6 +40,7 @@ module geod24.LocalRest;
 import vibe.data.json;
 
 static import C = std.concurrency;
+import std.datetime.systime : Clock, SysTime;
```
Comment (AndrejMitrovic) [inline source/geod24/LocalRest.d]: Move this where it's used?

**Comment by AndrejMitrovic** [comment]
Ok using a union for the messages. We can use a proper variant later, to prevent pulling in dependencies right now. (Rule of three, amirite?)

File: source/geod24/LocalRest.d [hunk 9]
```diff
@@ -400,11 +400,64 @@ public final class RemoteAPI (API) : API
 
     private static void spawned (Implementation) (CtorParams!Implementation cargs)
     {
+        import std.datetime.systime : Clock, SysTime;
+        import std.algorithm;
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Can you narrow down this import ?

File: source/geod24/LocalRest.d [hunk 10]
```diff
@@ -400,11 +400,64 @@ public final class RemoteAPI (API) : API
...
+            {
+                Response res;
+                Command cmd;
+            }
+
+            ubyte tag;
+        }
+
+        // can't embed SysTime in TimeCommand as it has internal aliasing
+        struct Sleep
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: What I had in mind was more of a `Control` struct where `filter` would be stored as well. But not a big deal.

File: source/geod24/LocalRest.d [hunk 11]
```diff
@@ -400,11 +400,64 @@ public final class RemoteAPI (API) : API
...
+                scheduler.spawn(() => handleCommand(arg, node, filter));
+            }
+            else static if (is(T == Response))
+            {
+                scheduler.pending = arg;
+                scheduler.waiting[arg.id].c.notify();
+            }
+        }
+
+        /// messages which are waiting to be handled due to a node.sleep() call
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: Extra `/` here

File: source/geod24/LocalRest.d [hunk 12]
```diff
@@ -400,11 +400,64 @@ public final class RemoteAPI (API) : API
...
+        // if the node is "sleeping", but doesn't want to drop messages,
+        // it should store them here. we need to track the queue of messages
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: That sentence gave me cancer.
```
Stores messages which are waiting to be handled, for example due to a `node.sleep()` call with `drop = false`.
We need to [...]

File: source/geod24/LocalRest.d [hunk 13]
```diff
@@ -400,11 +400,64 @@ public final class RemoteAPI (API) : API
...
+        // on our own, to avoid accidentally dropping priority messages in
```
Comment (Geod24) [inline source/geod24/LocalRest.d]: priority messages ?

**Review by Geod24 — APPROVED** [review]
Just a few nits otherwise LVGTM

**Comment by AndrejMitrovic** [comment]
Ok I fixed all that stuff.

**Review by Geod24 — APPROVED** [review]
I assume there's not good way to test this, as it is a race condition.
